share/wake/lib/system/io.wake

```diff
@@ -39,6 +39,35 @@ target writeImp inputs mode path content =
 
         makeRunner "write" (\_ Pass 0.0) pre post virtualRunner
 
+    # There are a couple likely bad paths that we don't want the user writing to
+    # so we give good error messages for these cases
+    require False = path ==* ""
+    else
+        failWithError
+        "Attempt to write to the path of the empty string. This was likely a mixup between the path and the content."
+
+    require False = path ==* "."
+    else failWithError "Attempt to write to write to the root workspace"
+
+    # We don't want `write` to write to anything outside of the workspace via
+    # a relative path
+    require False = matches `\.\..*` path
+    else failWithError "Attempt to write outside of the workspace"
+
+    # We don't want `write` to write to anything outside of the workspace via
+    # an absolute path
+    require False = matches `/.*` path
+    else failWithError "Attempt to write to an absolute path"
+
+    # Source files should never be deleted so we check for this case
+    def scan dir regexp = prim "sources"
```

Comment:
Calling `prim "sources"` for every call to `write` is expensive no? Do we cache the result somewhere?

Reply:
Wake reads the full list at the start and then does a linear regex scan through it. The linear scan could be improved to be a `Tree` in this case since we don't need a regex but this is the same cost we incur when we `source` a file today so I think its fine? I'll test it on a larger build to see what the effect is.